tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java

```diff
@@ -0,0 +1,227 @@
+
+    private final String service;
+    private final TChannel tchannel;
+    private final List<InetSocketAddress> routers;
+    private final long advertiseTimeout;
+    private final long advertiseInterval;
+
+    private Timer advertiseTimer = new Timer(true);
+
+    private HyperbahnClient(Builder builder) {
+        this.service = builder.service;
+        this.tchannel = builder.channel;
+        this.routers = builder.routers;
+        this.advertiseTimeout = builder.advertiseTimeout;
+        this.advertiseInterval = builder.advertiseInterval;
+    }
+
+    public ListenableFuture<Response<AdvertiseResponse>> advertise()
+        throws InterruptedException, TChannelError {
+
+        final AdvertiseRequest advertiseRequest = new AdvertiseRequest();
+        advertiseRequest.addService(service, 0);
+
+        // TODO: options for timeout, hard fail, etc.
+        final Request<AdvertiseRequest> request = new Request.Builder<>(
+            advertiseRequest,
+            HYPERBAHN_SERVICE_NAME,
+            HYPERBAHN_ADVERTISE_ENDPOINT
+        )
+            .setTTL(advertiseTimeout, TimeUnit.SECONDS)
+            .build();
+
+        // TODO: should be part of tchannel peer selection
+        final InetSocketAddress router = this.routers.get(new Random().nextInt(this.routers.size()));
+        ListenableFuture<Response<AdvertiseResponse>> responseFuture = null;
+        try {
+            responseFuture = this.tchannel.callJSON(
+                    router.getAddress(),
+                    router.getPort(),
+                    request,
+                    AdvertiseResponse.class
+            );
+        } catch (TChannelError ex) {
+            // TODO: should be moved into tchanne as retry ...
+            this.logger.error("Advertise failure: " + ex.toString());
+
+            // re-throw for now
+            // TODO: should really handle internally
+            throw ex;
+        }
+
+        responseFuture.addListener(new Runnable() {
+            @Override
+            public void run() {
+                if (destroyed.get()) {
+                    return;
+                }
+
+                scheduleAdvertise();
+            }
+        }, new Executor() {
+            @Override
+            public void execute(Runnable command) {
+                command.run();
+            }
+        });
+
+        return responseFuture;
+
+    }
+
+    private void scheduleAdvertise() {
+        if (destroyed.get()) {
+            return;
+        }
+
+        advertiseTimer.schedule(new TimerTask() {
+            @Override
+            public void run() {
+                try {
+                    ListenableFuture<Response<AdvertiseResponse>> responseFuture = advertise();
+                } catch (Exception ex) {
+                    logger.error(service + " failed to advertise. " + ex.getMessage());
+                }
+            }
+        }, advertiseInterval);
```

Comment:
AdvertiseInterval is fixed (1 min). Only retry on failure should back off. 
We don't really have retry on failure yet ...